File: frontend/src/services/dmca.ts
import api from "./api"

/** Mirrors ``chk_dmca_complaints_status`` and the Python ``DmcaComplaintStatus``. */
export type DmcaStatus = "received" | "upheld" | "rejected" | "withdrawn"

export const DMCA_STATUSES = {
  RECEIVED: "received",
  UPHELD: "upheld",
  REJECTED: "rejected",
  WITHDRAWN: "withdrawn",
} as const satisfies Record<string, DmcaStatus>

/** The three a person may choose. `received` is where a row starts and is
 *  not somewhere it can be put back: un-deciding a complaint would erase a
 *  strike an account closure may already rest on. */
export type DmcaDecision = Exclude<DmcaStatus, "received">

export interface DmcaComplaint {
  id: string
  received_at: string
  complainant_name: string
  complainant_email: string
  complainant_organization: string | null
  work_described: string
  material_location: string
  uploaded_by: string | null
  uploader_name: string | null
  uploader_notified_at: string | null
  status: DmcaStatus
  resolved_at: string | null
  resolved_by: string | null
  resolution_note: string | null
  /** Where this one sat in the uploader's sequence when it was upheld. */
  strike_number: number | null
  /** Where that person stands today — the number somebody deciding the next
   *  complaint actually needs. */
  uploader_upheld_count: number
  uploader_account_closed: boolean
}

/**
 * The complaints ledger.
 *
 * Admin-only on both ends: the rows name a complainant and a member of the
 * school, and § 512 asks for a record rather than for a public register.
 */
export const dmcaService = {
  async list(params?: { status?: DmcaStatus }): Promise<DmcaComplaint[]> {
    const { data } = await api.get<DmcaComplaint[]>("/admin/dmca/complaints", {
      params,
    })
    return data
  },

  /** Write down a notice that arrived by email. */
  async record(body: {
    complainant_name: string
    complainant_email: string
    complainant_organization?: string | null
    work_described: string
    material_location: string
    uploaded_by?: string | null
  }): Promise<DmcaComplaint> {
    const { data } = await api.post<DmcaComplaint>("/admin/dmca/complaints", body)
    return data
  },

  /**
   * Record what was decided. One-way: the server refuses a second decision on
   * the same complaint, because a strike may already have closed an account
   * and renumbering the history would leave the ledger unable to explain it.
   */
  async decide(
    id: string,
    body: {
      status: DmcaDecision
      resolution_note?: string | null
      uploader_notified?: boolean
    },
  ): Promise<DmcaComplaint> {
    const { data } = await api.patch<DmcaComplaint>(
      `/admin/dmca/complaints/${encodeURIComponent(id)}`,
      body,
    )
    return data
  },
}

File: backend/app/schemas/dmca.py
"""Shapes for the complaints ledger."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.schemas._request import RequestModel

#: Mirrors ``chk_dmca_complaints_status`` and ``DmcaComplaintStatus``.
DmcaStatusLiteral = Literal["received", "upheld", "rejected", "withdrawn"]

#: The statuses a human may set. ``received`` is where a row starts and is not
#: somewhere it can be put back: un-deciding a complaint would erase a strike
#: that an account closure may already have been based on.
DmcaDecisionLiteral = Literal["upheld", "rejected", "withdrawn"]


class DmcaComplaintCreate(RequestModel):
    """A notice as it arrived, transcribed by whoever opened the email.

    Deliberately not a public form. § 512(c)(3) wants a signature and a
    statement made under penalty of perjury, and a web form that collects
    neither produces something that looks like a notice and is not one. The
    published procedure says to write to the address on ``/dmca``; this is
    where that letter is written down.
    """

    complainant_name: str = Field(min_length=1, max_length=200)
    complainant_email: EmailStr
    complainant_organization: str | None = Field(default=None, max_length=200)
    #: 512(c)(3)(A)(ii) — the work claimed.
    work_described: str = Field(min_length=1, max_length=4000)
    #: 512(c)(3)(A)(iii) — where the material is.
    material_location: str = Field(min_length=1, max_length=2000)
    #: Who put it there, when it can be worked out.
    uploaded_by: UUID | None = None


class DmcaComplaintDecision(RequestModel):
    """What was decided, and whether the uploader has been told."""

    status: DmcaDecisionLiteral
    resolution_note: str | None = Field(default=None, max_length=4000)
    #: Setting this stamps "the uploader was told" at the current time. It is
    #: a flag rather than a timestamp because the only honest value is now:
    #: a date typed in by hand is a claim, not a record.
    uploader_notified: bool = False


class DmcaComplaintOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    received_at: datetime
    complainant_name: str
    complainant_email: str
    complainant_organization: str | None
    work_described: str
    material_location: str
    uploaded_by: UUID | None
    uploader_name: str | None = None
    uploader_notified_at: datetime | None
    status: DmcaStatusLiteral
    resolved_at: datetime | None
    resolved_by: UUID | None
    resolution_note: str | None
    strike_number: int | None
    #: How many upheld complaints this uploader has now, in total. The column
    #: says where this one sat in the sequence; this says where they stand
    #: today, which is the number somebody deciding the next one needs.
    uploader_upheld_count: int = 0
    #: Whether that count has passed the published threshold. Computed rather
    #: than stored so the answer cannot drift from the rule.
    uploader_account_closed: bool = False
